src/oauth-shim.js

```diff
@@ -196,12 +196,34 @@ oauth_shim.redirect = function(req, res, next) {
 
 		path += (hash ? '#' + param(hash) : '');
 
-		res.writeHead(302, {
-			'Access-Control-Allow-Origin': '*',
-			'Location': path
-		});
+		try {
+			res.writeHead(302, {
+				'Access-Control-Allow-Origin': '*',
+				'Location': path
+			});
+
+			res.end();
+		} catch(err) {
+			res.writeHead(500, {
+				'Access-Control-Allow-Origin': '*',
+			});
 
-		res.end();
+			if (err.message === "Invalid character in header content [\"Location\"]") {
+				// If the authenticating server returned a malformed response (e.g. an HTML page) that was accidentally
+				// parsed as form data, then we might have illegal characters in the oauthshim.data object's property names
+				// or values.
+				res.end(JSON.stringify({
+					error: err.message,
+					error_message: 'The authenticating server included illegal header characters in its response (it might not have responded with an expected format)',
+				}));
+			} else {
+				// For any other error, don't risk leaking sensitive data to the client. Just return an unknown error.
+				res.end(JSON.stringify({
+					error: 'Unknown Error',
+					error_message: 'Unknown Error',
+				}));
+			}
+		}
```

Comment:
This is the only change of consequence.

Note that I opted for a `500` error. I haven't really put in the work to know that the response from the OAuth server is invalid. It's just that we can't process it.